### PR TITLE
IGNITE-14299 .NET: Support BinaryArrays in services

### PR DESCRIPTION
**Pull request by nizhikov**
Thank you for submitting the pull request to the Apache Ignite.

In order to streamline the review of the contribution 
we ask you to ensure the following steps have been taken:

### The Contribution Checklist
- [ ] There is a single JIRA ticket related to the pull request. 
- [ ] The web-link to the pull request is attached to the JIRA ticket.
- [ ] The JIRA ticket has the _Patch Available_ state.
- [ ] The pull request body describes changes that have been made. 
The description explains _WHAT_ and _WHY_ was made instead of _HOW_.
- [ ] The pull request title is treated as the final commit message. 
The following pattern must be used: `IGNITE-XXXX Change summary` where `XXXX` - number of JIRA issue.
- [ ] A reviewer has been mentioned through the JIRA comments 
(see [the Maintainers list](https://cwiki.apache.org/confluence/display/IGNITE/How+to+Contribute#HowtoContribute-ReviewProcessandMaintainers)) 
- [ ] The pull request has been checked by the Teamcity Bot and 
the `green visa` attached to the JIRA ticket (see [TC.Bot: Check PR](https://mtcga.gridgain.com/prs.html))

### Notes
- [How to Contribute](https://cwiki.apache.org/confluence/display/IGNITE/How+to+Contribute)
- [Coding abbreviation rules](https://cwiki.apache.org/confluence/display/IGNITE/Abbreviation+Rules)
- [Coding Guidelines](https://cwiki.apache.org/confluence/display/IGNITE/Coding+Guidelines)
- [Apache Ignite Teamcity Bot](https://cwiki.apache.org/confluence/display/IGNITE/Apache+Ignite+Teamcity+Bot)

If you need any help, please email dev@ignite.apache.org or ask anу advice on http://asf.slack.com _#ignite_ channel.

### CODE REVIEW

**Comment by nizhikov** [comment]
Hello @ptupitsyn, can you, please, take a look at my changes?

File: modules/platforms/dotnet/Apache.Ignite.Core/Impl/Services/ServiceProxyInvoker.cs [hunk 1]
```diff
@@ -102,6 +102,12 @@ internal static class ServiceProxyInvoker
                         methodName, svcType, argsLength));
 
             // 2) There is more than 1 method with specified name - resolve with argument types.
+            methods = methods.Where(m => AreMethodArgsCompatible(arguments, m.GetParameters(), true)).ToArray();
```
Comment (ptupitsyn) [inline modules/platforms/dotnet/Apache.Ignite.Core/Impl/Services/ServiceProxyInvoker.cs:105]: I tried reverting all changes in this file, and all tests pass. Please double check - either we need more tests, or those changes are unnecessary.

**Review by ptupitsyn — CHANGES_REQUESTED** [review]

File: modules/platforms/dotnet/Apache.Ignite.Core/Impl/Services/ServiceProxyInvoker.cs (hunk 1 above)
Comment (nizhikov) [reply modules/platforms/dotnet/Apache.Ignite.Core/Impl/Services/ServiceProxyInvoker.cs:105]: `ServicesClientTestBinaryArrays#testOverloadResolution` failed without changes with exception: 
> Apache.Ignite.Core.Client.IgniteClientException : Failed to invoke proxy: there are 2 methods 'Foo' in type 
> 'Apache.Ignite.Core.Tests.Client.Services.TestServiceOverloads' with (Person[]) arguments, can't resolve ambiguity.

Comment (ptupitsyn) [reply modules/platforms/dotnet/Apache.Ignite.Core/Impl/Services/ServiceProxyInvoker.cs:105]: Now I understand, we attempt a more strict check to reduce ambiguity. 

Can you please rename `checkEq` to `strictTypeCheck` and invert the logic? Right now we perform equality check when `checkEq = false`, which is confusing.

Comment (ptupitsyn) [reply modules/platforms/dotnet/Apache.Ignite.Core/Impl/Services/ServiceProxyInvoker.cs:105]: And let's add a condition to make it clear that we try to resolve ambiguity here:

```suggestion
        if (methods.Length > 1)
            methods = methods.Where(m => AreMethodArgsCompatible(arguments, m.GetParameters(), true)).ToArray();
```

**Review by ptupitsyn — CHANGES_REQUESTED** [review]

**Comment by nizhikov** [comment]
@ptupitsyn All your current comments fixed. Please, take a look.

File: modules/platforms/dotnet/Apache.Ignite.Core/Impl/Services/ServiceProxyInvoker.cs (hunk 1 above)
Comment (nizhikov) [reply modules/platforms/dotnet/Apache.Ignite.Core/Impl/Services/ServiceProxyInvoker.cs:105]: Fixed.

**Review by ptupitsyn — APPROVED** [review]

**Comment by ptupitsyn** [comment]
@nizhikov looks good to me, thanks!

**Pull request merged**